File: golang/binary-tree-dfs/count_good_nodes_in_binary_tree.go
package main

func countGoodNodes(node *TreeNode, maxVal int) int {
	if node == nil {
		return 0
	}

	count := 0

	if node.Val >= maxVal {
		count = 1
	}

	newMax := max(maxVal, node.Val)

	count += countGoodNodes(node.Left, newMax)
	count += countGoodNodes(node.Right, newMax)

	return count
}

func goodNodes(root *TreeNode) int {
	return countGoodNodes(root, root.Val)
}

File: java/src/main/java/com/sanedge/interview/binary_tree_dfs/CountGoodNodesInBinaryTree.java
package com.sanedge.interview.binary_tree_dfs;

import com.sanedge.interview.utils.TreeNode;

public class CountGoodNodesInBinaryTree {

    public int countGoodNodes(TreeNode node, int maxVal) {
        if (node == null) {
            return 0;
        }

        int count = 0;

        if (node.val >= maxVal) {
            count = 1;
        }

        int newMax = Math.max(maxVal, node.val);

        count += countGoodNodes(node.left, newMax);
        count += countGoodNodes(node.right, newMax);

        return count;
    }

    public int goodNodes(TreeNode root) {
        return countGoodNodes(root, root.val);
    }
}
